packages/react/src/hooks/useContractRead.ts

```diff
@@ -0,0 +1,79 @@
+import { Contract, type Address, type JsonAbi, type Provider } from 'fuels';
+import type { FunctionNames, InputsForFunctionName } from 'src/types';
+
+import { useNamedQuery } from '../core/useNamedQuery';
+import { QUERY_KEYS } from '../utils';
+
+type ContractData<TAbi extends JsonAbi> = {
+  address: Address;
+  abi: TAbi;
+  provider: Provider;
+};
+
+type ContractReadProps<
+  TAbi extends JsonAbi,
+  TFunctionName extends FunctionNames<TAbi>,
+> = {
+  contract: Contract | ContractData<TAbi>;
+  functionName: TFunctionName;
+  args: InputsForFunctionName<TAbi, TFunctionName>;
+};
+
+export const useContractRead = <
+  TAbi extends JsonAbi,
+  TFunctionName extends FunctionNames<TAbi>,
+>({
+  functionName,
+  args,
+  contract: _contract,
+}: ContractReadProps<TAbi, TFunctionName>) => {
+  const isContractData = 'abi' in _contract && 'address' in _contract;
+  const { abi, address, provider } = (_contract as ContractData<TAbi>) ?? {};
+  const chainId = _contract?.provider?.getChainId();
+
+  return useNamedQuery('contractRead', {
+    queryKey: QUERY_KEYS.contract(
+      isContractData ? address?.toString() : _contract?.id?.toString(),
+      chainId,
+      args?.toString(),
+    ),
+    queryFn: async () => {
+      const isValid = isContractData
+        ? !!abi && !!address && !!provider
+        : !!_contract && 'provider' in _contract;
+
+      if (!isValid) {
+        throw new Error(
+          'Contract or address, abi and provider are required to read the contract',
+        );
+      }
+      if (typeof chainId !== 'number') {
+        throw new Error('ChainId is required to read the contract');
+      }
```

Comment:
we don't need to validate the `chainId` as 1-it comes from the provider 2- it will not be used further in the flow